File: test-utils/src/main/java/ru/virgil/test_utils/fluent_request/BodyStepJson.java
package ru.virgil.test_utils.fluent_request;

import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;

import javax.annotation.Nullable;
import java.util.List;

@RequiredArgsConstructor
public class BodyStepJson implements BodyStepJsonStart, BodyStep {

    private final RequestModel requestModel;
    private final Requester requester;
    private final ObjectMapper objectMapper;
    @Nullable
    private JavaType byteArrayJavaType;

    @Override
    public BodyStepJson send(Object dto) {
        requestModel.setRequestBody(dto);
        return this;
    }

    @Override
    public BodyStepJson receive(Class<?> responseClass, Class<?>... responseClasses) {
        JavaType javaType = objectMapper.getTypeFactory().constructParametricType(responseClass, responseClasses);
        requestModel.setResponseJavaType(javaType);
        return this;
    }

    @Override
    public BodyStepJson receiveAsBytes() {
        byteArrayJavaType = objectMapper.getTypeFactory().constructParametricType(List.class, Byte.class);
        requestModel.setResponseJavaType(byteArrayJavaType);
        return this;
    }

    @Override
    public BodyStepJson exchange(Object dto, Class<?> responseClass, Class<?>... responseClasses) {
        requestModel.setRequestBody(dto);
        JavaType javaType = objectMapper.getTypeFactory().constructParametricType(responseClass, responseClasses);
        requestModel.setResponseJavaType(javaType);
        return this;
    }

    @Override
    public FinalStep<?> and() throws Exception {
        if (requestModel.getResponseJavaType() == null) {
            return new FinalStepNoType(requestModel, requester);
        } else {
            if (requestModel.getResponseJavaType().equals(byteArrayJavaType)) {
                return new FinalStepByteType(requestModel, requester);
            } else {
                return new FinalStepJsonType<>(requestModel, requester, objectMapper);
            }
        }
    }
}

File: tools/src/main/kotlin/ru/virgil/spring_tools/tools/testing/fluent_request/BodyStepJson.kt
package ru.virgil.spring_tools.tools.testing.fluent_request

import com.fasterxml.jackson.databind.JavaType
import com.fasterxml.jackson.databind.ObjectMapper


class BodyStepJson(
    private val requestModel: RequestModel,
    private val requester: Requester,
    private val objectMapper: ObjectMapper,
) : BodyStepJsonStart, BodyStep {

    private var byteArrayJavaType: JavaType? = null

    override fun send(dto: Any): BodyStepJson {
        requestModel.requestBody = dto
        return this
    }

    override fun receive(responseClass: Class<*>, vararg responseClasses: Class<*>): BodyStepJson {
        val javaType = objectMapper.typeFactory.constructParametricType(responseClass, *responseClasses)
        requestModel.responseJavaType = (javaType)
        return this
    }

    override fun receiveAsBytes(): BodyStepJson {
        byteArrayJavaType =
            objectMapper.typeFactory.constructParametricType(MutableList::class.java, Byte::class.java)
        requestModel.responseJavaType = (byteArrayJavaType)
        return this
    }

    override fun exchange(dto: Any, responseClass: Class<*>, vararg responseClasses: Class<*>): BodyStepJson {
        requestModel.requestBody = (dto)
        val javaType = objectMapper.typeFactory.constructParametricType(responseClass, *responseClasses)
        requestModel.responseJavaType = (javaType)
        return this
    }

    @Throws(Exception::class)
    override fun and(): FinalStep<*> {
        return if (requestModel.responseJavaType == null) {
            FinalStepNoType(requestModel, requester)
        } else {
            if (requestModel.responseJavaType == byteArrayJavaType) {
                FinalStepByteType(requestModel, requester)
            } else {
                FinalStepJsonType<Any>(requestModel, requester, objectMapper)
            }
        }
    }
}
